### DISCUSSION TITLE
Generating new getters & Setters

### DISCUSSION THREAD

**Discussion by Amutongi19** [pyswmm/pyswmm#536] [Q&A] [answered]
I am trying to expand the Linkparams to include pipe geometry like length and diameter.  I was getting stuck at the toolkitapi file where the LinkParams are defined as I do not know how it reads the .inp file to take in this information.  I want to add parameters for length, pipe roughness, and diameter which are all in the .inp file but I cannot figure it out.

class LinkParams(Enum):
    offset1 = 0  # double
    offset2 = 1  # double
    q0 = 2  # double
    qLimit = 3  # double
    cLossInlet = 4  # double
    cLossOutlet = 5  # double
    cLossAvg = 6  # double
    seepRate = 7  # double

**Comment by bemcdonnell** [comment, marked as answer]
@Amutongi19, if you're using PySWMM, you probably don't need to dive into the C code anymore to accomplish those getters/setters.  We introduced a new features with PySWMM-2 that allows you to access any parameter to modify your base INP. 

https://pyswmm.github.io/pyswmm/reference/api/pyswmm.simulation.SimulationPreConfig.html

However, if you're using the base C library, can help you open up some of these features.

**Reply by Amutongi19** [reply]
Thank you! This is perfect

**Reply by Amutongi19** [reply]
Is there a way to use this just as a getter?  Like just to retrieve certain inp parameters.

**Reply by bemcdonnell** [reply]
@Amutongi19 I'd give you 1000 high-fives if you implemented that inside SimulationPreConfig!  It would certainly be useful.

**Reply by bemcdonnell** [reply]
@Amutongi19 We usually recommend SWMMIO for that for now.  The longer-term plan is to merge together some of the features of pyswmm and swmmio to make it more seamless. Just need someone to offer to write the code!

**Reply by Amutongi19** [reply]
Yeah I'm just using SWMMIO in the interim.  I suppose I'll look into the C library if I can get time.

**Comment by joshipra** [comment]
Hello :)
For some reason, I cannot import SimulationPreConfig from pyswmm. Is there an additional step that needs to be done? 
Many thanks in advance...

**Reply by bemcdonnell** [reply]
What version of pyswmm are you using? This is a newer feature. Might need an upgrade

**Reply by joshipra** [reply]
I, indeed, had to upgrade to version 2.0.1, but it still did not work. The SystemSeries class, for example, does not work

**Reply by bemcdonnell** [reply]
Can you paste some of your code here?

**Reply by joshipra** [reply]
![image](https://github.com/pyswmm/pyswmm/assets/174731545/8a55e1b5-197b-4dd1-aa30-46e293d6e82f)
I could use SystemSeries this way but have not tried SimulationPreConfig yet

**Reply by bemcdonnell** [reply]
Here is how it is imported in the unit test 

https://github.com/pyswmm/pyswmm/blob/master/pyswmm/tests/test_simulation.py

**Reply by joshipra** [reply]
Thank you. It seems to be working. May I ask if we can use SimulationPreConfig to change LID soil parameters (porosity, for instance) during the simulation using SimulationPreConfig (as it says it is a "simple way to programmatically adjust nearly all model parameters")? Or is that still not allowed?

**Reply by bemcdonnell** [reply]
@joshipra It really depends on which parameters you are hoping to change.  PreConfig just changes the INP before the simulation.  As for the list of LID parameters open to modify during the simulation, these are well documented.  For example: 

<img width="1092" alt="image" src="https://github.com/pyswmm/pyswmm/assets/6105134/085bcc48-d9db-48aa-ad2d-99b1b9f79320">

many of these values are turned off since it would required a re-validation of the asset.  If they were opened, it could produce many unintended impacts to results.  That being said, if you have a use-case and research need for it, we could coach you along to make the modifications to SWMM.  What are your thoughts?

**Reply by joshipra** [reply]
Thank you for your quick response :). I am trying to model poor maintenance cases in which the bioretention cell's (just an example) soil parameters (say porosity or hydraulic conductivity) change over time. I noted a 'Clogging Factor' parameter for the storage layer but not for the soil layer, which would have been ideal for my case. Do you think this can be addressed? Many thanks in advance :)

**Reply by bemcdonnell** [reply]
@joshipra This sounds like a pretty good use-case.  Here is the API code that limits the setting behaviors:

https://github.com/pyswmm/Stormwater-Management-Model/blob/459db1d4dfc61ff994ae01f92eae64e378e08915/src/solver/toolkit.c#L1716

Why not have a look at this and see how you might modify it and how you would test the behavior inside of some unit tests.  I am supportive if you want to give it a go!  @karosc , thoughts?
